### PR TITLE
Add --keep option to allow to generate newsfile, but keep newsfragmen…

### LINKED ISSUE #129: Add option `--no`
For batch-processing (e.g. when testing), an option `--no` would be helpful:
```
--no           Do not ask for confirmation to remove news fragments, assume "no".
```

### PR DESCRIPTION
**Pull request by fizyk**
…ts - closes #129

# Description
Resolves issue #129 It's not a --no option, but given the generality of the --no option, --keep is a bit more descriptive than proposed --no.

# Checklist

* [x] Make sure changes are covered by existing or new tests.
* [x] For at least one Python version, make sure local test run is green.
* [x] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.
* [ ] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [ ] Ensure `docs/tutorial.rst` is still up-to-date.
* [x] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [ ] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

### CODE REVIEW

**Comment by adiroiban** [comment]
Hi Grzegorz. Many thanks for the nice PR.

I don't really understand why this is needed.

There is the "draft" command which is expected to generate the release notes, without removing the fragments.

It is true that right now, the draft command generates the result in the standard output.

But maybe a better way to fix this is to add a file option to the existing "draft" command.

This is just a suggestion.

**Comment by fizyk** [comment]
@adiroiban I do not have strong opinion on where. It seemed natural, although the --draft could also be the solution.

The underlying issue suggests it would be useful to generate a `current` changelog for development documentation, my issue is that having numerous fragment types I want to have a general changelog with all the changes and additionally specific categories being generated into its own, separate changelog file as well.

**Comment by adiroiban** [comment]
Thanks for the explanation. I think '--keep' makes sense.

> The underlying issue suggests it would be useful to generate a current changelog for development documentation, 

I think that this is "draft" use case.
But "draft" has clear documentation that it sends the result to stdout... so... it's complicated.

I will check the PR now and I hope we can merge it.

Thanks again

File: src/towncrier/_git.py
Comment (adiroiban) [inline src/towncrier/_git.py]: Can we can "answer_yes == True" here ?

I was expecting that this is validated in the caller code.
By handling this branch, somebody might think that this is a valid use-case.

```suggestion
    if answer_keep:
        # Not proceeding with the removal of the files.
        return
```

File: src/towncrier/build.py
Comment (adiroiban) [inline src/towncrier/build.py]: Maybe just make it a private method, not to encourage its usage outside of towncrier.

I thing that the missing docs is ok.
I can underestand the code even without docs.

```suggestion
def _validate_answer(ctx: Context, param: Option, value: bool) -> bool:
```

File: src/towncrier/build.py
Comment (adiroiban) [inline src/towncrier/build.py]: Can we have more help text here?

In an ideal case, the whole info from the "cli.rst" file would be placed here.

In this way, we can use cli-man to generate a man version of the docs...and we can also use `towncrier build --help` to get that info.

these days, when I use a CLI, in order to get some documentation, I first try my luck with "--help" and only if it fails I use "man" or the project website.

File: src/towncrier/test/test_build.py
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: I think that is very important to have a docstring for the test to describe its scope and purpose

Later we might have many overlaping tests and it will be hard to know which tests is for which purpose.

```suggestion
    def test_keep_fragments(self):
        """
        The `--keep` option will build the full final news file
        without deleting the fragment files and without
        any extra CLI interaction or confirmation.
        """
        runner = CliRunner()
```

File: src/towncrier/test/test_build.py
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: Maybe also have an assertion before calling the CLI that the NEWS file doesn't exists.

```suggestion
            # The NEWS file is created.
            # So this is not just `--draft`.
            self.assertTrue(os.path.isfile("NEWS.rst"))
```

File: src/towncrier/test/test_build.py
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: Can you please add a description to this test.

I am not sure why you are calling `build` twice

We don't have tests for the test code, so a docstring for the tests function is the only think that we have to assert that the test is valid.

maybe also assert that the news file is not created and the fragment files are not removed.

```suggestion
    def test_yes_keep_error(self):
        """
        It will fail to perform any action when the
        conflicting --keep and --yes options are provided.
        """
        runner = CliRunner()
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
This looks very good. Many thanks for writing the docs and tests.

It helps a lot with the review.

I left a few minor comments.

The biggest comment is the lack of docstring for the new tests.
Please use it to describe the targeted end-user functionality.

The towncrier project don't really has a good knnowledge of what it should do and how it should do.
Towncrier is pretty flexible and various people are using it for their use cases... and without a docstring for a tests, it's hard to know what is a side effect and what is an intended feature.

------------

I understand the use case, and I think that this is specific enough not to be included in the tutorial.

-----

> If you add new configuration options (or change the meaning of existing ones), make sure docs/configuration.rst reflects those changes.

@hynek  I think that here we should say

```diff
-If you add new configuration options
+If you add new template options
```
This PR has only CLI options so I think that we are fine.

File: src/towncrier/_git.py
Comment (fizyk) [reply src/towncrier/_git.py]: Hmm.... you're right

File: src/towncrier/test/test_build.py
Comment (fizyk) [reply src/towncrier/test/test_build.py]: I call it twice because the click triggers a callback in the order it parses the command line, not in the order in which they are defined in the command. That's why I also added the validator to both `--yes` and `--keep` 

I'll add the comment to test as well.

File: docs/cli.rst
Comment (adiroiban) [inline docs/cli.rst]: I think the important part here is that fragments are not removed.

I don't know if "Useful for automated tasks" adds any new info on top of "Don't ask for confirmation".
If it doesn't ask, it should be good for automation.

I don't know what to day about the specific use case.
From this one sentence, I don't fully undestand what "Where you have to generate several different news fragments for different parties." should do.

If we want to have this information in the docs,
I think we should create a separate doc page with use case scenarios.

Something similar to "How to keep changelong in markdown"

In this way we can add more info.

I guess that you are using towncrier to generate a "Release-Notes-Everyting" and a separate "Release-Notes-Bugs" and "Release-Notes-Security" and "Release-Notes-Features"... but it is not clear.

```suggestion
   Don't delete news fragments after the build and don't ask for confirmation whether to delete or keep the fragments.

```

File: src/towncrier/_git.py
Comment (adiroiban) [inline src/towncrier/_git.py]: Just a very minor comment. We can merge with this.
Do we need to be verbose here ?

I guess that it can help with troubleshooting to detect which files are used for the build process... so we can keep it.

We can just return ASAP and remove_files is a NOOP for `answer_keep = Yes`

**Review by adiroiban — APPROVED** [review]
Thanks. Changes looks good.
Only very minor comments that should not be a blocker.

Somehow, I have messed up the coverage reporting back to GitHub PR.
The coverage is ok, but posting back the comment fails.

I will try to fix `trunk` and then we can merge this.

Many thanks again for your help!

File: src/towncrier/_git.py
Comment (fizyk) [reply src/towncrier/_git.py]: Hmm....  This could kill two birds with one stone, not sure how would you feel about it though

```python
def remove_files(
    fragment_filenames: list[str], answer_yes: bool, answer_keep: bool
) -> None:
    if not fragment_filenames:
        return

    try:
        if answer_yes:
            click.echo("Removing the following files:")
        elif answer_keep:
            click.echo("Keeping the following files:")
            # Not proceeding with the removal of the files.
            return
        else:
            click.echo("I want to remove the following files:")
    finally:
        # Will always be printed, even for answer_keep to help with possible troubleshooting
        for filename in fragment_filenames:
            click.echo(filename)

    if answer_yes or click.confirm("Is it okay if I remove those files?", default=True):
        call(["git", "rm", "--quiet"] + fragment_filenames)
```

Comment (adiroiban) [reply src/towncrier/_git.py]: Thanks for the updates.

Usually, I don't like clever code :)

But if you like it, I am ok to merge something like this... maybe with an extra comment.
But I think the code is not that hard to read.
I don't see any good reason why not to have this code approved.

Maybe remove "elif". I as general rule, I try to code with as little else/elif as possible.

```python
def remove_files(
    fragment_filenames: list[str], answer_yes: bool, answer_keep: bool
) -> None:
    if not fragment_filenames:
        return

    try:
        if answer_keep:
            # Not proceeding with the removal of the files.
            click.echo("Keeping the following files:")
            return  # See the logic from finally.

        if answer_yes:
            click.echo("Removing the following files:")
        else:
            click.echo("I want to remove the following files:")
    finally:
        # Always show the source files,
        # even for answer_keep to help with possible troubleshooting
        for filename in fragment_filenames:
            click.echo(filename)

    if answer_yes or click.confirm("Is it okay if I remove those files?", default=True):
        call(["git", "rm", "--quiet"] + fragment_filenames)
```

**Comment by adiroiban** [comment]
I/we still need to fix the coverage reporting issue.

I did an initial investigation, and it looks like forks only get read permission

I can see that for the main repo, the permissions are `write` ... for example for the latest trunk merge

https://github.com/twisted/towncrier/actions/runs/3404794371/jobs/5662304592#step:1:17

But for this PR the permissions are only `read`

https://github.com/twisted/towncrier/actions/runs/3563302652/jobs/5985987960#step:1:17


@hynek @altendky Do you any a quick fix for standalone coverage reporting on forks ?

Thanks!

**Comment by hynek** [comment]
Maybe https://github.com/py-cov-action/python-coverage-comment-action? I think actions have different token handling

File: src/towncrier/test/test_build.py
Comment (hynek) [inline src/towncrier/test/test_build.py]: Could you use the with_isolated_runner decorator in new code please?

https://github.com/twisted/towncrier/blob/trunk/src/towncrier/test/helpers.py#L23

File: src/towncrier/test/test_build.py
Comment (hynek) [inline src/towncrier/test/test_build.py]: Same as above

**Review by hynek — CHANGES_REQUESTED** [review]

**Comment by adiroiban** [comment]
>  I think actions have different token handling

As far as I know, the actions are just a "step" , so they run in the context of the job.

We are using the default token...and the default token only has read permissions.
[As documented here](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)

There is this setting [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-the-default-github_token-permissions)  
but I don't know what are the side effects

File: src/towncrier/test/test_build.py
Comment (fizyk) [reply src/towncrier/test/test_build.py]: done :)

Comment (hynek) [reply src/towncrier/test/test_build.py]: Thanks! :)

**Comment by hynek** [comment]
@adiroiban sadly I'm rather clueless at all these things. If you want to just quickly unblock contributions I can offer [my method](https://hynek.me/articles/ditch-codecov-python/) that doesn't comment, but at least works for now. People just have to look into the CI output for missing lines.

**Comment by adiroiban** [comment]
Thanks @hynek . Yes, the commit status might be the only "safe" option.

In theory, we could give write access to forks, since our workflow has explicit permissions... but if in the future a new workflow is created without explicit permissions, this can lead to a security void, as we might accidentally give full access to the repo to forks.

At the same time, I think that new contributors need to have their first PR manually approved, so the workflow code is not executed right away, unattended.

I will look to create a new PR that will always send the commit status and will post a comment only for non-forks.

File: src/towncrier/_git.py
Comment (hynek) [inline src/towncrier/_git.py]: so, generally speaking I'm not a fan of this shell-style Python (that towncrier is full of and not specific to this PR – I'm just trying to have less of it :))

So while we're at refactoring: could we split this up into a function that does _not_ live within `_git.py` & returns a list of files that should be removed and minimizing `remove_files` to something resembling:

```python
def remove_files(files: list[str]) -> None
    if files:
        call(["git", "rm", "--quiet"] + files)
```

Comment (fizyk) [reply src/towncrier/_git.py]: sure :)

Comment (fizyk) [reply src/towncrier/_git.py]: done :)

File: src/towncrier/_remover.py
Comment (hynek) [inline src/towncrier/_remover.py]: Close, but not quite what I had in mind.

We haven't really achieved anything by this refactoring, because this function still deletes the files which makes it an action with pretty heavy side-effects.

My suggestion was that this function returns the `fragment_filenames` as a list and the caller calls `_git.remove_files()` on it unconditionally.

---

Secondarily, you've moved this function into a separate module, so you can easily monkeypatch it without affecting the CLI entry point – correct?

That's pretty heavy "test damage".

How about going a bit functional here:

```python
_git.remove_files(
    find_news_fragment_files_for_removal(  # <-- better name for remove_news_fragment_files after changes
        fragment_filenames, answer_yes, answer_keep,
        confirm_fn=partial(click.confirm, "Is it okay if I remove those files?", default=True)
    )
)
```

In the test you would simply pass `confirm_fn=lambda: True` or `confirm_fn=lambda: False` – look ma no patching!

I would usually not suggest something like that on a public repo, but if the alternative is adding a new module for just one function that is called in one place, it seems worth it – would you agree?

File: src/towncrier/newsfragments/129.feature
Comment (hynek) [inline src/towncrier/newsfragments/129.feature]:
```suggestion
Added ``--keep`` option to the ``build`` command that allows to generate a newsfile, but keeps the newsfragments in place.
This option can not be used together with ``--yes``.
```

File: src/towncrier/_remover.py
Comment (fizyk) [reply src/towncrier/_remover.py]: Hmmm.... you're right. I got lost a bit between reading your suggestion and sitting down to implement it.

---

When I look at it, the intermediate function only makes a decision on whether to remove the files or not. And would return the fragments_list or not. 
I wonder if it wouldn't be even better if we had something like this:

```python
if should_remove_news_fragment(fragment_filenames, answer_yes, answer_keep):
    _git.remove_files(fragment_files)
```

Then both functions here could be tested separately. And... still, the should_remove_news_fragments could receive confirm_fn partial...

---

Sorry, I tend to split functions sooner, especially if the original module is already several hundred lines big. I'll move it back since... you're right, it's not called anywhere else, and that's a perfect rule to keep 🤔 .

Comment (fizyk) [reply src/towncrier/_remover.py]: @hynek I don't think the functional approach will help in regards to the changed test.

The test that got the mock updated is an integration one (tests from all the way up by calling click command) and... I moved the click calls from `_git.remove` so unless I'll keep the intermediate within `_git` module, you'd have to change the mock anyway 🤔 

And to actually make use of the functional approach the test itself would have to be a unit one.
I'm not sure how to approach it.
And the last question, shall I move the intermediate function into the build module?

Comment (hynek) [reply src/towncrier/_remover.py]: > Sorry, I tend to split functions sooner, especially if the original module is already several hundred lines big.

yeah the CLI entry points need to be refactored HARD, but let's not blow this up even further 😇

---

since the list is never changed, agreed on `should_remove_news_fragments`

---

ok this click.confirm business is annoying. I was gonna suggest an abstraction, but I've noticed that it's literally only used in that one place so I'm gonna say to KISS and patch away on build.py without the confirm callable, unless you'd like to add more *unit*tests.

Comment (fizyk) [reply src/towncrier/_remover.py]: > but let's not blow this up even further

whew 😌 

---
> unless you'd like to add more unittests

I can't think of any that the integration tests wouldn't already cover... so not at this time. 
I'll try to remember next time I make a contribution though :)

**Review by hynek — APPROVED** [review]
sorry for the delay, I've been sick all week.

i think given the circumstances this is good to go.

**Comment by adiroiban** [comment]
Many thanks @fizyk for the changes and @hynek for the review.
I was busy at work pushing a review before the holiday break.

I think we can also look at pushing a new release for towncrier... to get this feature and also get latest documentation

**Comment by hynek** [comment]
Agreed – having the news docs would be great!

**Comment by fizyk** [comment]
@hynek @adiroiban thank you both for the review, discussion and satisfactory work :)
And I'm glad I could also close one issue from someone's issue board :)

**Pull request merged**
